File: overlay/clojure/hydra-kernel/src/main/clojure/hydra/overlay/clojure/lib/pairs.clj
(ns hydra.overlay.clojure.lib.pairs)

;; Pairs are 2-element lists: (list a b)

;; bimap :: (a -> c) -> (b -> d) -> Pair a b -> Pair c d
(def hydra_overlay_clojure_lib_pairs_bimap
  "Map over both elements of a pair."
  (fn [f] (fn [g] (fn [p] (list (f (first p)) (g (second p)))))))

;; first :: Pair a b -> a
(def hydra_overlay_clojure_lib_pairs_first
  "Get the first element of a pair."
  (fn [p] (first p)))

;; pair :: a -> b -> Pair a b
(def hydra_overlay_clojure_lib_pairs_pair
  "Construct a pair from two values."
  (fn [x] (fn [y] (list x y))))

;; second :: Pair a b -> b
(def hydra_overlay_clojure_lib_pairs_second
  "Get the second element of a pair."
  (fn [p] (second p)))

File: overlay/go/hydra-kernel/src/main/go/hydra/overlay/go/lib/pairs/pairs.go
// Package pairs implements the hydra.lib.pairs primitives.
package pairs

import "hydra.dev/hydra/overlay/go/util"

// Bimap : (a -> c) -> (b -> d) -> pair<a,b> -> pair<c,d>
func Bimap(f any, g any, p any) any {
	pr := p.(util.Pair)
	return util.NewPair(f.(func(any) any)(pr.First), g.(func(any) any)(pr.Second))
}

// First : pair<a,b> -> a
func First(p any) any { return p.(util.Pair).First }

// Pair : a -> b -> pair<a,b>. The pair constructor.
func Pair(a any, b any) any { return util.NewPair(a, b) }

// Second : pair<a,b> -> b
func Second(p any) any { return p.(util.Pair).Second }
